js/store.js

```diff
@@ -112,51 +216,139 @@ const store = {
+  },
+
+  // check if third party is on the whitelist (owned by the first party)
+  // returns true if it is and false otherwise
+  onWhiteList(firstPartyFromRequest, thirdPartyFromRequest) {
+    if (thirdPartyFromRequest && this.firstPartyWhiteList) {
+      const hostnameVariantsFirstParty
+        = this.getHostnameVariants(firstPartyFromRequest);
+      for (let i = 0; i < hostnameVariantsFirstParty.length; i++) {
+        if (this.firstPartyWhiteList
+          .hasOwnProperty(hostnameVariantsFirstParty[i])) {
+          // first party is in the whitelist
+          const index = this.firstPartyWhiteList[hostnameVariantsFirstParty[i]];
+          const hostnameVariantsThirdParty
+            = this.getHostnameVariants(thirdPartyFromRequest);
+          for (let j = 0; j < hostnameVariantsThirdParty.length; j++) {
+            if (this.thirdPartyWhiteList[index]
+              .includes(hostnameVariantsThirdParty[j])) {
+              return true;
+            }
+          }
+          return false;
+        }
+      }
+    }
+    return false;
   },
 
   setFirstParty(hostname, data) {
     if (!hostname) {
       throw new Error('setFirstParty requires a valid hostname argument');
     }
+    const isNewWebsite = this.isNewWebsite(hostname);
+
     this.setWebsite(hostname, data);
+
+    if (isNewWebsite) {
+      this.updateChild(this.outputWebsite(hostname, this._websites[hostname]));
+    }
   },
 
+  // @todo use data from capture to update UI/filter graph
+  // eslint-disable-next-line no-unused-vars
   setThirdParty(origin, target, data) {
-    let newThirdParty = false;
+    let isNewWebsite = false;
     if (!origin) {
       throw new Error('setThirdParty requires a valid origin argument');
     }
 
-    const firstParty = this.getWebsite(origin);
-    const thirdParty = this.getWebsite(target);
+    let firstParty = this.getWebsite(origin);
+    let thirdParty = this.getWebsite(target);
 
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
     if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
+      // third party is not currently linked to first party
+      ({ firstParty, thirdParty, isNewWebsite }
+        = this.whiteListCheck(origin, target, firstParty, thirdParty));
     }
-    if (!('firstPartyHostnames' in firstParty)) {
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
     }
 
-    this.setWebsite(origin, firstParty);
+    this.setFirstParty(origin, firstParty);
     this.setWebsite(target, thirdParty);
 
-    if (newThirdParty) {
-      this.updateChild(this.outputWebsite(target, data, origin));
+    if (isNewWebsite && thirdParty['isVisible']) {
```

Comment:
1. I intentionally wasn't setting the 1st party in this function for clarity, instead just presence check it or we should think about initing it elsewhere.

Either:
```
this.getFirstParty(origin); // this in tern will call get website and initialise it to have all the data we care about
```
Or:
```
if (!('thirdPartyHostnames' in  firstParty)
    || !firstParty['thirdPartyHostnames'].includes(target)) { 
```

2. yup 👍 I said it was untested and you were the tester haha

3. this logic is wrong, you are overriding abc.xyz and google.co.uk with the data from mozilla.org when all 3 load googleanalytics.com This is why I explicitly was only getting in this other function. The other first party get is just to check if the link exists between mozilla.org and GA. This is irrelevant mostly to the data we set in the other function.

4. Nah I flat out missed:
```
    if (!('isVisible' in thirdParty)
        || !thirdParty.isVisible) { 
      ...
   } else {
     this.addFirstPartyRelationship(origin, target);
   }
```